README.md
# EinsatzMonitor-Alarm-PDF-Converter

## Dokumentation

*Toolset für den [EinsatzMonitor der Feuer Software GmbH](https://feuersoftware.com/einsatzmonitor/).*  
*Bei Fragen erreicht ihr mich in dem ihr ein `Issue` erstellt oder per Mail [`dev@hecht.space`](mailto:dev@hecht.space?subject=[EMAPC]%20Deine%20Frage).*

Der EMAPC (EinsatzMonitor-Alarm-PDF-Converter) bietet ein Toolset zur Konvertierung einer eingehenden Alarm-PDF in ein Format, das vom EinsatzMonitor lesbar ist.
Dieses Tool extrahiert die relevanten Informationen der Alarm-PDF in eine schlüssel-wert-formatierte Datei, die durch die Mustererkennung des EinsatzMonitors gelesen werden kann.  

### Mitarbeit

Möchtest du mit an emapc arbeiten? Zögere nicht, Fehler zu melden oder Feature-Anfragen zu stellen. Bitte nutze dazu die Issues-Funktion.  
Du kannst mir auch gerne bei der Implementierung dieses Tools helfen. Erstelle hierzu ein fork diese Repos -> Arbeite auf diesem Fork und stelle mir dann einen PR ein.  
Stelle sicher, dass du den aktuellen Status des `main`-Branches verwendest, bevor du einen Merge anforderst.  
Halte dich bei den commit messages bitte an die [conventional commits](https://www.conventionalcommits.org/de/v1.0.0/).

### Du willst EMAPC verwenden?

Kein Problem. Es gibt keine Einschränkungen für die Verwendung dieses Projekts.  
Wenn du etwas Feedback hast: Bitte erstellen ein Issue und füge das Label `Feedback` hinzu.  
Wenn du noch etwas anderes zu diesem Projekt sagen willst: Erstelle ein Issue und füge das Label `Hi` hinzu.

### INSTALLATION

EMAPC kann als Kommandozeilentool (cli-tool) oder als Dienst (service) betrieben werden.
Ersteres wird zur einmaligen Verarbeitung der Eingabe verwendet. Der Dienst hingegen läuft, bis es explizit gestoppt wird. Er überwacht das angegebene Verzeichnis und verarbeitet jede Datei, die in das Verzeichnis gelegt wird.

**Downloads:**

- Windows Dienst Installer [64-bit](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/releases/latest/download/emapc-windows-service-installer-x64.msi)<!-- #ASSET#0 -->
- Windows Dienst Installer [32-bit](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/releases/latest/download/emapc-windows-service-installer-x86.msi)<!-- #ASSET#1 -->
- Linux Installer (cli & Dienst) [64-bit](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/releases/latest/download/emapc-installer-linuxx64.sh)<!-- #ASSET#2 -->
- Linux Installer (cli & Dienst) [32-bit](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/releases/latest/download/emapc-installer-linuxx86.sh)<!-- #ASSET#3 -->
- Weitere Downloads (wie die Binärdateien können auf der [Releaseseite](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/releases/latest) gefunden werden)

Emapc unterstützt Linux (32-Bit [x86] und 64-Bit [x64]) und Windows (32-Bit [x86] und 64-Bit [x64]).  
In dem Installationsprogramm für Windows (.msi-Datei) sind alle Abhängigkeiten enthalten. Beim Ausführen des Skripts werden diese im Installationsordner (`C:\Program Files\EMAPC`) abgelegt. Hierzu gehört auch die Konfigurationsdatei (`C:\Program Files\EMAPC\emapc.conf.yml`). Die Beispielkonfigurationsdatei (`C:\Program Files\EMAPC\example.conf.yml`) sollte nicht verändert werden um bei updates die neue Standardkonfiguration zu erhalten. Weiter wird der `Einsatzmonitor Alarm PDF Converter`-Dienst in den Windows Diensten installiert und kann hier gestoppt und gestartet werden (zu finden unter: `Win + R` -> services.msc -> `ENTER`). Dieser Dienst startet nach Neustarts oder Crashes des Computers automatisch. EMAPC kann in den Windows `Programme und Features` deinstalliert werden. (Der Windows Installer unterstützt zur Zeit nur den Dienst, nicht das Kommandozeilentool).

Das Installationsskript für Linux lädt auch ein Deinstallationsskript herunter. Es befindet sich ebenfalls in `/usr/local/emapc`. Das Skript deinstalliert und entfernt alle Komponenten des Kommandozeilen-Tools und des Dienstes.

### VERWENDUNG

#### Dienst

Der Dienst verwendet eine Konfigurationsdatei, die im Arbeitsverzeichnis abgelegt wird und den Namen `emapc.conf.yml` hat. Der Installer legt das Arbeitsverzeichnis auf `/usr/local/emapc/` bzw. `C:\Program Files\EMAPC` fest. Der Installer lädt außerdem eine Standardkonfiguration.  
Die Konfiguration wird [unten](#konfiguration) erklärt.

> **HINWEIS:** Da emapc Standard-Betriebssystem-Dienste verwendet, können Standard-Betriebssystem-Tools verwendt werden, um diesen Dienst zu verwalten. Auf Windows kann der Dienstmanager verwendet werden. Auf Windows verwendet emapc nssm als Wrapper, so dass auch nssm-Befehle verwendet werden können. [Weitere Informationen](http://nssm.cc/). Der Installer enthält auch die ausführbare nssm-Datei (zu finden unter `/usr/local/emapc/` oder `C:\Program Files\EMAPC`).

##### Verwendung mit dem Einsatzmonitor

Für die Verwendung mit dem Einsatzmonitor müssen ein paar Einstellungen im Einsatzmonitor gemacht werden. Alle der folgenden Einstellungen befinden sich in den Auswertungs-Einstellungen (Menü unten links -> Auswertung).

0. EinsatzMonitor-Alarm-PDF-Converter (EMAPC) und der Einsatzmonitor sollte schon installiert sein.
1. Um die Alarme über E-Mail zu empfangen muss unter E-Mail der entsprechende Mail-Server eingerichtet werden. Zusätzlich müssen "Download Anhang" und "Nur Anhang" aktiviert werden und unter "Anhang Speicherort" der Ordner `C:\Program Files\EMAPC\input` angegeben werden.
2. Als nächstes muss unter "File Überwachung" der Schalter umgelegt werden. Nun sollte der Einsatzmonitor neugestartet werden.
3. Jetzt muss die Konfiguration von EMAPC angepasst werden. Notwendig ist hierbei lediglich die Anpassung des Ausgabeordners (`service -> output`). Dieser muss zu `/Users/<Nutzername>/Einsatz_Monitor/Text_Input/` geändert werden (Nutzername sollte der Nutzer des Einsatzmonitor sein).
4. EMAPC gibt eine Textdatei mit einem festen Format (in der Konfiguration definiertem) aus. Dieses Format muss nun unter "Pattern" eingepflegt werden (Beispiel unten).
5. Letztlich muss unter "AAO-Auswertung" der Separator ";" und das Pattern "{{FUNKRUFNAME}}" eingetragen werden.

Pattern:

```txt
Meldebild:{{EINSATZSTICHWORT}}
Ort:{{ORT}}
Ortsteil:{{ORTSTEIL}}
Straße:{{STRASSE}}
Ortszusatz:{{ZUSATZFELD2:Zusatzinformation}}
Sondersignal:{{ZUSATZFELD1:Sondersignal}}
Bemerkung:{{ZUSATZFELD2:Bemerkung}}
Stichwort 2:{{ZUSATZFELD1:Stichwort}}
Stichwort:{{ZUSATZFELD3:Bemerkung}}
Einsatzmittel:{{AAO}}
```

#### CLI

`emapc --config konfigurationsDateiPfad eingabeDateiOderOrdner ausgabeDateiOderOrdner`

##### NOTWENDIGE ARGUMENTE

- **eingabeDateiOderOrdner:**  
  Dies muss ein gültiger Pfad zu einer PDF-Datei sein. (Die Alarm-PDF-Datei).  
  Oder dies kann ein Pfad zu einem Ordner sein. In diesem Fall wird emapc alle PDF-Dateien in diesem Ordner verarbeiten.

- **ausgabeDateiOderOrdner:**  
  Dies kann der Pfad zu einer *(noch nicht existierenden)* Ausgabedatei sein.  
  > In diesem Fall verwendet emapc den angegebenen Dateinamen.  

  **\- ODER -**  
  Dies kann der Pfad zu einem bestehenden Verzeichnis sein.  
  > In diesem Fall übernimmt emapc den Dateinamen aus der Eingabe.

- **--config** *<Pfad/zur/Konfiguration/Datei.json>*  
  Diesem Parameter sollte ein gültiger Pfad zu einer benutzerdefinierten Konfigurationsdatei folgen.  

### Wie es funktioniert

Unabhängig davon, ob das Kommandozeilentool oder den Dienst verwendet wird, wird emapc durch eine Konfigurationsdatei konfiguriert. In dieser kann angeben werden, welche Eingaben erwartet werden, was in der Ausgabe stehen soll und einige zusätzliche Dinge. Die Konfiguration wird [später](#konfiguration) erklärt.
Wie oben erwähnt verwendet emapc PDF-Dateien als Eingabe. Das Tool iteriert über alle Dateien, die als Input angegeben wurden (entweder als Argument oder in der Konfiguration).  
Nach einiger Initialisierung (wie dem Parsen der Eingabeargumente und der Konfiguration) wird jede Eingabedatei einzeln verarbeitet.  
EMAPC verwendet das `pdftotext` Kommandozeilen-Tool des [xpdf Projekts](https://www.xpdfreader.com/). Dieses Tool ist im Installer enthalten, so dass emapc vollständig `battery-included` ist (alle Abhängigkeiten werden mitgeliefert).  
Nun veratbeitet emapc diesen Text. Zuerst versucht emapc, einige spezielle Schlüssel (genannt `inText`-Schlüssel) zu extrahieren. Diese Schlüssel werden in der Konfiguration festgelegt, wie [unten](#konfiguration) beschrieben. Das Werkzeug sucht nach Zeilen, die einen der Spezialschlüssel enthalten. Für jeden Spezialschlüssel werden alle gefundenen Zeilen verkettet und gespeichert (und später bei der Ausgabe verwendet). Die gefundenen Zeilen werden in den nächsten Schritten nicht mehr beachtet.  
Emapc wird die nächsten Schritte für jeden Abschnitt (in der Konfiguration angegeben) einzeln durchführen.  
Der erste dieser Schritte besteht darin, festzustellen, wo die Abschnitte beginnen. Hierfür wird nach Zeilen mit dem angegebenen Schlüsselwort gesucht. Für Abschnitte, die sich über mehrere Seiten erstrecken, gibt es mehrere Zeilen mit dem Schlüsselwort. Emapc behandelt sie als Unterabschnitte und führt sie später zusammen.  
Das Ende eines (Unter-)Abschnitts wird durch eine Leerzeile (Zeile mit nur Leerzeichen, Tabulatoren oder einer leeren Zeichenfolge) erkannt.  
Nun werden Abschnitte, die ein tabellenartiges Format haben, anders verarbeitet als Abschnitte, die einen Schlüsselwert ähnliches Format haben (das Format wird in der Konfiguration angegeben).  
Aber zuerst versucht emapc, das Format des Abschnitts zu erraten, sofern der Typ des Abschnitts auf `try` gesetzt ist. Emapc rät, indem es schaut, ob die Zeile mit dem Schlüsselwort mit einigen Leerzeichen (Leerzeichen oder Tabulatoren) beginnt. Wenn dies der Fall ist, wirg angenommen, dass der Abschnitt eine Tabelle ist, andernfalls wird angenommen, dass es sich um ein Schlüsselwertformatähnlichen Abschnitt handelt *(vielleicht sollte dieser Test anders gemacht werden)*.  

Zu der Verarbeitung der Schlüsselwert-Abschnitte: Emapc durchläuft die Zeilen dieses Abschnitts in einer Schleife. Für jede Zeile wird der Schlüssel extrahiert, indem nach dem ersten Vorkommen von zwei Leerzeichen hintereinander gesucht und dann den Text davor genommen. Anschließend wird der Wert in der gleichen Zeile extrahiert. Dazu nimmt emapc den Text zwischen den ersten beiden Leerzeichen aufeinander folgenden Leerzeichen und dem zweiten Vorkommen davon. Dieser Text ist der extrahierte Wert. Der gesamte Text, der sich nach dem zweiten Auftreten von zwei aufeinanderfolgenden Leerzeichen befindet, wird später als 'restlicher Text' behandelt. Jetzt behandelt emapc die Zeilen unterhalb der aktuellen Zeile. Erst versucht emapc zu erkennen, ob der Text in diesen Zeilen Teil des aktuellen Wertes sein sollte. Hierzu wird geprüft, ob diese Zeilen mit einigen Leerzeichen beginnen (Es werden alle folgenden Zeilen angeschaut, bis eine mit einem neuen Schlüssel gefunden wird). Die gefundenen Werte werden verkettet. Schließlich betrachtet emapc den (oben erwähnten) restlichen Text. Hierbei schaut emapc, ob dieser Text mit einem ähnlichen Format wie der aktuelle Schlüssel beginnt oder andere bekannte Schlüsselwörter (wie 'Datum:') verwendet. Wenn ja hängt emapc diesen Text als separate Schlüsselwerte an.

Zu den tabellenähnlichen Abschnitten: Hier schaut emapc auf die erste Zeile. Es wird angenommen, dass in dieser Zeile die Spaltenüberschriften stehen. Emapc vesucht den Text der jeweiligen Kopfzeile und den zugehörigen Startindex zu extrahieren. Es wird davon ausgegangen, dass die jeweilige Spalte in allen Zeilen am gleichen Startindex (+- 2) beginnt. Weiter wird davon ausgegangen, dass die Spaltenüberschriften durch mindestens zwei Leerzeichen getrennt sind. Falls die erste Spalte keine Überschrift enthält (z.B. bei Tabellen mit Zeilenüberschriften), fügt emapc eine leere Überschrift und den dazu gehörigen Index (0) ein. Nach dem Extrahieren der Spaltenüberschriften iteriert emapc über die folgenden Zeilen und versucht, die Werte jeder Spalte für die jeweilige Zeile durch den erkannten Spaltenstartindex und den Startindex der nächsten Spalte zu extrahieren.

Emapc schreibt nun das Ergebnis in eine Textdatei im Schlüsselwertformat.  
Zum Schreiben iteriert emapc über alle Ausgabeschlüssel (festgesetzt in [configuration](#configuration)). Dies stellt sicher, dass die Ausgabedatei für jeden Durchlauf das gleiche Format hat (natürlich ändern sich die Daten für verschiedene Eingaben).  
Die Ausgabe wird in einem schlüsselwertähnlichen Format geschrieben. Jede Zeile enthält ein neues Schlüsselwertpaar. Der Schlüssel und der Wert werden mit dem konfigurierten keyValueSeparator (Standard: `: `) getrennt.  
Wenn es sich bei dem Schlüssel um einen Schlüssel handelt, der mehrere Werte enthält, werden alle Werte in dieser Zeile ausgegeben, getrennt durch das konfigurierte Trennzeichen (Voreinstellung: `;`).  
Optional kann emapc ein Präfix und ein Suffix zu einem Wert hinzufügen (ein Suffix und/oder ein Präfix pro Schlüssel-Wert-Paar).  
Beispiel:

```txt
# ein Wert
Meldebild=TH Ölspur
# mehrere Werte
Einsatzmittel=FW1/1;FW1/2
```

**WICHTIG:** Wenn irgendein Schritt der Konvertierung fehlschlägt, schreibt empac den Rohtext der Eingabe-PDF in die Ausgabe-Textdatei!

### Konfiguration

Für emapc ist die Konfiguration erforderlich. Die Konfiguration wird in einer [yaml](https://de.wikipedia.org/wiki/YAML)-Datei vorgenommen. Eine Beispielkonfiguration ist [hier](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/blob/main/emapc.conf.yml) zu finden (Das Beispiel sollte für Feuerwehren im Landkreis Biberach funktionieren).  

Die Konfiguration gliedert sich grundsätzlich in 3 Teile: `Input`, `Output` und einen optionalen `Service` Teil. Der Dienst-Teil (`service`) ist erforderlich, wenn emapc als Dienst/Service betrieben wird.  

Der Input-Teil sagt emapc, wie die Eingabedateien in etwa aussehen werden. Dazu werden die erwarteten (Eingabe-)Abschnitte und spezielle 'inTextKeys' angegeben. Die inTextKeys sind eine Liste von speziellen Schlüssel (sie werden [oben](#wie-es-funktioniert) beschrieben). Die Abschnitts-Konfiguration andererseits ist eine Abbildung von Sektionsnamen (sollte das Sektions-Schlüsselwort/Header aus der Eingabedatei sein) auf den Sektionstyp. Erlaubte Typen sind `keyValue`, `table` und `try`, falls emapc versuchen sollte, den Typ zu erkennen (beschrieben [oben](#wie-es-funktioniert)).  
*Beispiel:*

```yaml
input:
  sections:
    Einsatzanlass: keyValue
    Einsatzort: try
    EM: table
  inTextKeys:
    - Sondersignal
```


CHANGELOG.md
# [1.3.0-next.1](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/compare/v1.2.2-next.1...v1.3.0-next.1) (2021-01-30)


### Bug Fixes

* behebt Probleme mit der geschriebenen Datei im Fallback modus ([aa0cf4c](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/commit/aa0cf4ca9b0572b12b2e832a82311d8be818bcea))
* fixt rest string wenn value leer ist ([0a2ada0](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/commit/0a2ada0ef9c07e851dea7c3acf0cdf0f87a1f4e4))


### Features

* es können nun standartwerte angegeben werden ([70730ed](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/commit/70730edfe43729df102a15f6ef307683c3744d89))

## [1.2.2-next.1](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/compare/v1.2.1...v1.2.2-next.1) (2021-01-26)


### Bug Fixes

* meldung über grafische Oberfläche wenn der Eingabeordner gelöscht wird ([33b59a3](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/commit/33b59a3346e4ba345ba3b4c1463bc67016523420))
* **config:** der Dienst lädt nach einer Konfigurationsänderung neu ([967c7cd](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/commit/967c7cd97c6d2fccca0ea69b23044fd42d2d050b))
* **config:** unterstützt windows pfade besser ([51a827d](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/commit/51a827d16cff60f1904571972eb5a130cda3fee6))

## [1.2.2-next.1](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/compare/v1.2.1...v1.2.2-next.1) (2021-01-25)


### Bug Fixes

* meldung über grafische Oberfläche wenn der Eingabeordner gelöscht wird ([33b59a3](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/commit/33b59a3346e4ba345ba3b4c1463bc67016523420))
* **config:** der Dienst lädt nach einer Konfigurationsänderung neu ([967c7cd](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/commit/967c7cd97c6d2fccca0ea69b23044fd42d2d050b))
* **config:** unterstützt windows pfade besser ([51a827d](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/commit/51a827d16cff60f1904571972eb5a130cda3fee6))

## [1.2.2-next.1](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/compare/v1.2.1...v1.2.2-next.1) (2021-01-24)


### Bug Fixes

* meldung über grafische Oberfläche wenn der Eingabeordner gelöscht wird ([33b59a3](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/commit/33b59a3346e4ba345ba3b4c1463bc67016523420))
* **config:** der Dienst lädt nach einer Konfigurationsänderung neu ([967c7cd](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/commit/967c7cd97c6d2fccca0ea69b23044fd42d2d050b))
* **config:** unterstützt windows pfade besser ([51a827d](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/commit/51a827d16cff60f1904571972eb5a130cda3fee6))

## [1.2.1](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/compare/v1.2.0...v1.2.1) (2021-01-23)


### Bug Fixes

* **installer:** removes security warning by signing installer ([2fb208a](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/commit/2fb208a6ae299208cb2afe779a001e32a89ba66f))

# [1.2.0](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/compare/v1.1.4...v1.2.0) (2021-01-22)


### Features

* **installer:** besserer Windows Installer ([83cf402](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/commit/83cf40292b2127d1cf649f719c974c9f10af22bf))

## [1.1.4](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/compare/v1.1.3...v1.1.4) (2021-01-19)


### Bug Fixes

* **dependencies:** fixes missing module error ([48241d1](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/commit/48241d1fd0c89c68b616b225a371ff02021b07f4))

## [1.1.3](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/compare/v1.1.2...v1.1.3) (2021-01-15)


### Bug Fixes

* **extractor:** fixes wrong output on empty input values ([965e684](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/commit/965e68414979ddda7aad2341ba26318bff7be66d)), closes [#23](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/issues/23)
* fixes cli ([713cc62](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/commit/713cc62444111caa112208a880eba1dcac5c7d00))

## [1.1.2](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/compare/v1.1.1...v1.1.2) (2020-10-28)


### Bug Fixes

* fixes uninstaller ([bab1035](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/commit/bab1035adc5ff34a40efdbab3d7c8392af4a0a7d))

## [1.1.1](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/compare/v1.1.0...v1.1.1) (2020-10-27)


### Bug Fixes

* adds uninstaller ([bdca29a](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/commit/bdca29ac2fd472d26c5223dc91ba9373dd0b2b75))

# [1.1.0](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/compare/v1.0.0...v1.1.0) (2020-10-25)


### Bug Fixes

* **installer:** configuration will not be overwritten anymore ([73413a4](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/commit/73413a4c9d8c55604935d7129049768f7be74341))


### Features

* **output:** adds fallback if some required keys are empty ([086b9cd](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/commit/086b9cd328695fb6c229e5d3cfb2d67822c9430a))

# 1.0.0 (2020-10-24)


### Bug Fixes

* **extractor:** fixes failure if section key word is contained elsewhere in the file ([701f402](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/commit/701f40246c2f6b4bc7afb43af0b1e8c06a324623))
* **extractor:** fixes wrong handling of similar keys in restString ([0ae0317](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/commit/0ae031711066dfb0d5859b687d6c71d579da917e))
* fixes build ([64ea8e9](https://github.com/matzehecht/EinsatzMonitor-Alarm-PDF-Converter/commit/64ea8e94efcdc79931ceffd22246d7d388839c60))

src/index.ts
import * as TraceIt from 'trace-it';
import { existsSync, promises as fs } from 'fs';
import * as os from 'os';
import { promisify } from 'util';
import * as child from 'child_process';
import { Input, Output } from './config';
import * as path from 'path';
import * as Extractor from './extractor';
import * as utils from './utils';
import { KeyValueKey, ListByWordKey, ValueByWordKey, ValueIndexKey } from './config';
import { Writable } from 'stream';

export async function convert(
  inputFileOrDir: string,
  isInputDir: boolean,
  outputFileOrDir: string,
  inputConfig: Input,
  outputConfig: Output,
  parentTransaction?: TraceIt.Transaction
) {
  const transaction = parentTransaction?.startChild('convert');

  const proms: Promise<void>[] = [];

  if (!isInputDir) {
    const fileChild = transaction?.startChild('file');
    const fnWithoutExt = path.basename(inputFileOrDir, '.pdf');
    const outputFile = path.extname(outputFileOrDir) === '' ? path.join(outputFileOrDir, `${fnWithoutExt}.txt`) : outputFileOrDir;

    if (!existsSync(path.dirname(outputFile)) || !(await fs.stat(path.dirname(outputFile))).isDirectory()) {
      await fs.mkdir(path.dirname(outputFile));
    }

    if (path.extname(inputFileOrDir) !== '.pdf') {
      utils.logInfo('INPUT', `skipping the file ${path.resolve(inputFileOrDir)} as it is not a pdf file!`);
    } else {
      utils.logInfo('INPUT', `processing ${path.resolve(inputFileOrDir)}!`);

      proms.push(run(inputConfig, outputConfig, inputFileOrDir, outputFile, fileChild));
    }
    fileChild?.end();
  } else {
    const files = await fs.readdir(inputFileOrDir, { withFileTypes: true });

    if (!existsSync(outputFileOrDir) || !(await fs.stat(outputFileOrDir)).isDirectory()) {
      await fs.mkdir(outputFileOrDir);
    }

    files.every((file) => {
      const fileChild = transaction?.startChild('file');
      const filename = file.name;
      const fnWithoutExt = path.basename(filename, path.extname(filename));

      if (path.extname(filename) !== '.pdf') {
        utils.logInfo('INPUT', `skipping the file ${path.resolve(filename)} as it is not a pdf file!`);
      } else {
        utils.logInfo('INPUT', `processing ${path.resolve(filename)}!`);

        proms.push(run(inputConfig, outputConfig, path.join(inputFileOrDir, filename), path.join(outputFileOrDir, `${fnWithoutExt}.txt`), fileChild));
      }
      fileChild?.end();
      // return true;
    });
  }

  transaction?.end();
  return Promise.all(proms);
}

function toWriteString(text: string) {
  return `${text}${os.EOL}`;
}

function joinSafe(array: string[], separator?: string): string {
  return array.map((cur: string) => (!separator ? cur : cur.replace(RegExp(separator, 'gi'), separator === ';' ? ',' : ';'))).join(separator);
}

async function run(inputConfig: Input, outputConfig: Output, inputFile: string, outputFile: string, transaction?: TraceIt.Transaction) {
  const pdftotext = getBinary();
  const pdftotextArgs = ['-simple', inputFile, '-'];
  utils.logInfo('Read pdf', `command: ${pdftotext}, args: ${pdftotextArgs}`);
  const readChild = transaction?.startChild('read');
  const raw = (await promisify(child.execFile)(pdftotext, pdftotextArgs, { encoding: 'latin1' })).stdout;
  readChild?.end();

  try {
    const extractChild = transaction?.startChild('extract');
    const output = Extractor.extract(raw, inputConfig, extractChild);
    extractChild?.end();

    const writeChild = transaction?.startChild('write');
    const writer = new Writer();

    const separator = outputConfig.separator || ';';
    const keyValueSeparator = outputConfig.keyValueSeparator || ': ';
    const outputKeys = outputConfig.keys;

    const entries = Object.entries(outputKeys);

    for (const [key, keyConfig] of entries) {
      if (!output[keyConfig.inputSection]) {
        throw new OutputError(`section ${keyConfig.inputSection} not configured in input`);
      }

      const sectionValues = output[keyConfig.inputSection];

      if (Array.isArray(sectionValues)) {
        // is table section
        if ('type' in keyConfig) {
          // is ListByWordKey
          const thisKeyConfig = keyConfig as ListByWordKey;

          if (thisKeyConfig.type === 'column') {
            const values = sectionValues
              .map((row) => row[thisKeyConfig.inputKeyWord])
              .filter((value) => {
                return !thisKeyConfig.filter || !value.match(RegExp(thisKeyConfig.filter));
              });

            if (keyConfig.required && values.length === 0) throw new OutputError(`key ${key} is required but empty`);

            const valueToWrite = values.length !== 0 ? values : keyConfig.default ? keyConfig.default : [];

            writer.write(
              toWriteString(`${key}${keyValueSeparator}${thisKeyConfig.prefix || ''}${joinSafe(valueToWrite, separator)}${thisKeyConfig.suffix || ''}`)
            );
          } else {
            const row = sectionValues.find((row) => Object.values(row)[0] === thisKeyConfig.inputKeyWord);
            const values = row ? Object.values(row) : undefined;
            const curatedValues = values?.slice(1).filter((value) => !thisKeyConfig.filter || !value.match(RegExp(thisKeyConfig.filter))) || [];

            if (keyConfig.required && curatedValues.length === 0) throw new OutputError(`key ${key} is required but empty`);

            const valueToWrite = curatedValues.length !== 0 ? curatedValues : keyConfig.default ? keyConfig.default : [];

            writer.write(
              toWriteString(`${key}${keyValueSeparator}${thisKeyConfig.prefix || ''}${joinSafe(valueToWrite, separator)}${thisKeyConfig.suffix || ''}`)
            );
          }
        } else if ('index' in keyConfig) {
          // is ValueByWordKey
          const thisKeyConfig = keyConfig as ValueByWordKey;
          const rows = sectionValues.filter((row) => thisKeyConfig.inputKeyWords.includes(Object.values(row)[0]));
          const values = rows
            .map((row) => Object.values(row)[thisKeyConfig.index + 1])
            .filter((value) => !thisKeyConfig.filter || !value.match(RegExp(thisKeyConfig.filter)));

          if (keyConfig.required && values.length === 0) throw new OutputError(`key ${key} is required but empty`);

          const valueToWrite = values.length !== 0 ? values.join(' ') : keyConfig.default ? keyConfig.default : '';

          writer.write(toWriteString(`${key}${keyValueSeparator}${thisKeyConfig.prefix || ''}${valueToWrite}${thisKeyConfig.suffix || ''}`));
        } else if ('rowIndex' in keyConfig) {
          // is ValueIndexKey
          const thisKeyConfig = keyConfig as ValueIndexKey;

          const row = sectionValues[thisKeyConfig.rowIndex];
          const value = Object.values(row)[thisKeyConfig.columnIndex].trim();

          if (keyConfig.required && value.length === 0) throw new OutputError(`key ${key} is required but empty`);

          if (thisKeyConfig.filter && value.match(RegExp(thisKeyConfig.filter))) {
            writer.write(toWriteString(`${key}${keyValueSeparator}`));
          } else {
            writer.write(
              toWriteString(`${key}${keyValueSeparator}${thisKeyConfig.prefix || ''}${value || keyConfig.default || ''}${thisKeyConfig.suffix || ''}`)
            );
          }
        } else {
          throw new OutputError(`key ${key} should be configured as Table Key`);
        }
      } else if (Object.keys(sectionValues).length !== 0) {
        // is keyValue section
        if (!('inputKeyWords' in keyConfig)) {
          throw new OutputError(`key ${key} should has inputKeyWords as it is a key value key`);
        }
        const thisKeyConfig = keyConfig as KeyValueKey;

        const value = thisKeyConfig.inputKeyWords
          .reduce((prev, key) => {
            const thisValue = sectionValues[key];
            if (!thisValue) return prev;
            if (thisKeyConfig.filter && thisValue.match(RegExp(thisKeyConfig.filter))) return prev;
            return [prev, thisValue].join(' ');
          }, '')
          .trim();

        if (keyConfig.required && value.length === 0) throw new OutputError(`key ${key} is required but empty`);

        writer.write(toWriteString(`${key}${keyValueSeparator}${thisKeyConfig.prefix || ''}${value || keyConfig.default || ''}${thisKeyConfig.suffix || ''}`));
      } else {
        writer.write(toWriteString(`${key}${keyValueSeparator}${keyConfig.default || ''}`));

        if (keyConfig.required) throw new OutputError(`key ${key} is required but empty`);

        utils.logInfo('OUTPUT', 'empty key', key);
      }
    }

    writeChild?.end();
    await fs.writeFile(outputFile, writer.toString());
  } catch (err) {
    await fs.writeFile(outputFile, raw);
    throw err;
  }
}

function getBinary() {
  let script = 'pdftotext';
  let pathToScript = '';

  switch (os.arch()) {
    case 'x64':
      script += '64';
      break;
    case 'x32':
      script += '32';
      break;
    default:
      throw new Error(`Architecture ${os.arch()} not supported`);
  }

  switch (os.platform()) {
    case 'darwin':
      pathToScript = 'mac';
      break;
    case 'linux':
      pathToScript = 'linux';
      break;
    case 'win32':
      pathToScript = 'win';
      break;
    default:
      throw new Error(`OS ${os.platform()} not supported`);
  }

  return path.join(utils.basePath, `./lib/pdftotext/${pathToScript}/${script}`);
}

export class OutputError extends Error {
  constructor(m: string) {
    super('OUTPUT - ' + m);
    Object.setPrototypeOf(this, OutputError.prototype);
  }
}

class Writer extends Writable {
  private data: string = '';
  _write(chunk: string, enc: string, cb: () => void) {
    this.data += chunk;
    cb();
  }
  toString() {
    return this.data;
  }
}
